warehouse/accounts/views.py

```diff
@@ -800,9 +800,38 @@ def request_password_reset(request, _form_class=RequestPasswordResetForm):
         if user is None:
             user = user_service.get_user_by_email(form.username_or_email.data)
         if user is not None:
-            email = first_true(
-                user.emails, pred=lambda e: e.email == form.username_or_email.data
+            verified_email = first_true(
+                user.emails,
+                pred=lambda e: e.email == form.username_or_email.data and e.verified,
             )
+
+            if not verified_email:
+                # If the user is found but the email is not verified,
+                # tell the user to contact support instead
+                if unverified_email := first_true(
+                    user.emails,
+                    pred=lambda e: e.email == form.username_or_email.data,
+                ):
```

Comment:
This conditional doesn't work when attempting to request password reset by username: `verified_email` becomes `None`, but this query is searching for an email that matches the username, not the email address, so it will never succeed.

I'm surprised the tests & coverage are OK with this? Something is falling between the cracks here.

Reply:
The comparison isn't changed much from before - before verified, we were looking for the first true user.emails match based on the inbound form, so I'm not sure how it worked before.

Reply:
I hunted it down, and it's super subtle down here:
https://github.com/pypi/warehouse/blob/eb5e8fb62e48763f46868cc0fde6577f71cfb57c/warehouse/email/__init__.py#L96-L99

So if we never pass an email, it'll go to the primary on file.

Reply:
Which might be an unverified email, yeah?

Reply:
Currently, yes. Which is the behavior I'm trying to disallow.

Reply:
OK, but right now because you've set `allow_unverified=True`, when a user requests a password reset by username , and they don't have any verified emails on their account, nothing happens. They get the usual "If you submitted a valid username or email address, an email has been sent to your registered email address" but nothing happens: they have submitted a valid username but we won't send any emails.